adr/ADR-1.md

```diff
@@ -4,152 +4,280 @@
+```
+
+Will list the stream names starting at number 1024.
+
+## Anatomy of a Response
+
+Generally we have a few forms of response with some representative ones shown here:
+
+### Good response
 
 ```json
 {
-  "type": "io.nats.jetstream.api.v1.stream_info",
-  "time": "2020-04-23T16:51:18.516363Z",
-  "config": {
-    "name": "STREAM",
-    "subjects": [
-      "js.in"
-    ],
-    "retention": "limits",
-    "max_consumers": -1,
-    "max_msgs": -1,
-    "max_bytes": -1,
-    "max_age": 31536000,
-    "max_msg_size": -1,
-    "storage": "file",
-    "num_replicas": 1
-  },
-  "state": {
-    "messages": 95563,
-    "bytes": 40104315,
-    "first_seq": 34,
-    "last_seq": 95596,
-    "consumer_count": 1
-  }
+  "type": "io.nats.jetstream.api.v1.stream_names_response",
+  "total": 1,
+  "offset": 0,
+  "limit": 1024,
+  "streams": [
+    "KV_NATS"
+  ]
 }
 ```
 
-Consumer Info Error:
+This is not an error response (no `error` field), it's paged (has `total`, `offset` and `limit`) and is of the type `io.nats.jetstream.api.v1.stream_names_response` which indicates its schema (see below).
+
+### Error Response
 
 ```json
 {
-  "type": "io.nats.jetstream.api.v1.consumer_info",
+  "type": "io.nats.jetstream.api.v1.consumer_info_response",
   "error": {
-    "description": "consumer not found",
     "code": 404,
-    "error_code": 10059
+    "err_code": 10059,
+    "description": "stream not found"
   }
 }
 ```
 
-Here we have a minimally correct response with the additional error object.
+This is a response to a consumer info request, it's type is `io.nats.jetstream.api.v1.consumer_info_response` which indicates its schema (see below).
+This is an error response, where the `description` field is variable and can change it's content between server versions. The fields of a healthy response are not shown for an error.
 
-In the `error` struct we have `description` as a short human friendly explanation that should include enough context to
-identify what Stream or Consumer acted on and whatever else we feel will help the user while not sharing privileged account
-information.  These strings are not part of the API promises, we can update and re-word or translate them at any time. Programmatic
-error handling should look at the `code` which will be HTTP like, 4xx human error, 5xx server error etc. Finally, the `error_code`
-indicates the specific reason for the 404 - here `10059` means the stream did not exist, helping developers identify the
-real underlying cause. 
+## Schemas
 
-More information about the `error_code` system can be found in [ADR-7](0007-error-codes.md).
+All requests and responses that are in JSON format have JSON Schema Draft 7 documents describing them.
 
-Ideally the error response includes a minimally valid body of what was requested but this can be very hard to implement correctly.
+For example the request to create a stream has the type `io.nats.jetstream.api.v1.stream_create_request` and the response
```

Comment:
Should library written now look for `v1.` in the response and fail if not found? Meaning is that expected that if it one day becomes `v2.` it means that the client will not be able to understand the response?

Reply:
Some people believe we'll never see version 2 - we're only adding fields ever.  

Generally I check the full type matches exactly one that I have a known type for - though granted I've fully incorporated 100% validated requests and responses in the API and is probably more than is realistic.

In the CLI/Terraform/GitHub Actions any response that does not validate 100% of does not have a known schema would cause a panic and I also check requests to the server against the schemas.

The real answer is somewhere in the middle - I'd check the full type field personally at least.

Reply:
So to make sure I understand, if I send a "stream_create_request", then when the library gets a response, it should fail if the response's type is not "io.nats.jetstream.api.v1.stream_create_response"?

Reply:
I fail in that case yes, is it the 100% right thing to do? Not sure, I think its a very strong signal that things are in a bad way somewhere and someone should know.  So that's what the CLI does.

Clients might want to be a bit more relaxed about it, seems risky? But if you are getting a response today without that type I can guarantee you things are very broken.

Reply:
It would be interesting to see what clients have checked the response type. Java and .NET have not. If this is important, its relatively trivial to do but it's work to be done. Should I add a line to the parity sheet and and open an issue?

Reply:
For today I dont think it's a problem - we're way off from having a v2 - hopefully never.  I wouldnt loose sleep on it, we can punt till we get to v2.

Reply:
For the CLI I would via validation also detect if I am running against a server returning new fields and it would fail. 

I am not sure what clients can do as we have no negotiation around api versions etc. These fields are the only guide - and we have had some breaking changes already. 

I don’t know what the answer is for languages who can’t do schemas tbh :)